README

A whole set of small utilities around automake/conf and co. by
Jean-Daniel Pauget.

Most of them are still badly finished, with hardcoded path and
other ugglynesses the same kind.


bozoconf:
    tests if any "./configure" script lays in the current dir and try to
    run it. If there is such script, try to use "autoall" (see below) for
    creating one and running it.

    "bozoconf --version" cleverly dumps the version and revision of most tools
    involved in the autobuild process, useful for bug tracking.
    
autoall:
    runs one after the other the bestof gnu automake autoconf suite in
    the proper order (or at least proper enough for the author's projects)

repozclean:
    tries to clean most files created by the previous scripts in order to
    have a cleaned source tree ready for repository comparisons, updates and check-ins.

cvsstat:
    recurse into current source tree and check via cvs for needed updates or local
    changes. It should also detect missing files/directories. OBSOLETE.

maketags:
    recurses into current source tree and in a set of hardcoded headers' trees so
    that a huge concatenation of "ctags" file is generated.

vimall:
    recurses into current source tree and launch a vi (vim) editing the whole set
    of sources, sorted by most recent time-stamps. It uses the above "maketags"
    for easying the navigation.
    
cptobackup:
    creates a timestamped-named copy of the given file or directory. It echoes
    the choosen command that was applied during the process for the user's
    understanding. It should not overwrite any existing file, a rather add some
    counting digits to the timestamped name.
    For directories, ownership and permissions are kept by using the gnu extension
    "cp -a" of the unix "cp" command.
    For files, the backuped permissions (with time-stamps) should fall under the
    "umask" and "cp" rules set when the operation is performed. If any doubt, double
    check the result with "ls -l"...

mvtobackup:
    the same as the above "cptobackup" except that files or directories are moved
    with the unix "mv" command instead of being copied.

for first time installation without ./configure, here's a rought canvas :
	(g)libtoolize --copy && aclocal && autoheader && automake   --add-missing && autoconf
	./configure
	make
	make install



ChangeLog
2013-12-10: jd         : rev-1-2-9
                         macosX macports support
2013-10-20: jd         : rev-1-2-7
                         hcpp compliance added
                         .cc files are considered secondary and yet left over
                         in maketags
2010-04-09: jd         : rev-1-2-0
                         debian packaging
2005-02-08: jd         : rev-1-1-1
                         cleanup in regard with build out of directory
			 badly named "clean4cvs" renamed in "repozclean"
			 better handling of (truncated, though) long lists
			 of sources (vimdiff and maketags)
2005-01-07: jd         : rev-1-1-0
                         consistency cleanup in regard with cross-compilation
			 badly named "configure" renamed in "bozoconf"
2004-05-04: jd         : cvs tag rev-0-9-13 .
                         some more tunings in regard with new autotools
                         better global consistency
2004-04-29: jd         : handling etags added
                         --track-changes added
			 humanised clean4cvs
2004-04-27: jd         : cvs tag rev-0-9-11 .
                       : cvs tag rev-0-9-10 .
                       : whole set of corrections with newer autotools
                       : whole set of corrections in regard with msys/mingw
                       : addition of rev, previous functionnality presence missing
2004-04-22: jd         : cvs tag rev-0-9-9 .
                       : moved to *.in precursor for all scripts
2004-04-21: jd         : added --version flag in almost all scripts
2004-03-28: jd         : cvs tag rev-0-9-8 .
                       : the commands are packed together, bozomake's born


src/count.c
/* 
 * $Id$
 * Bozomake Copyright (C) 2002,2003 Cristelle Barillon & Jean-Daniel Pauget
 * A whole set of building utilities
 *
 * bozomake@disjunkt.com  -  http://bozomake.disjunkt.com/
 * 
 * This file is part of Bozomake.
 * 
 * Loopsoids is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Loopsoids is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Loopsoids; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * you can also try the web at http://www.gnu.org/
 */
#include <stdio.h>
#include <stdlib.h>

int main (int nb, char ** argv)
{
    long order = 1;
    int ordered = 1;
    char buf1 [4096],
	 buf2 [4096],
	 *p, *q;
    p = buf1, q = buf2;
    *p = 0;

    {	int i;
	for (i=1 ; i<nb ; i++)
	{   if (*argv[i] == '-')
	    switch (argv[i][1])
	    {	case '0':
		    order = 0;
		    break;
		default:
		    break;
	    }
	}
    }

    while (fgets (q, 4096, stdin) != NULL)
    {	
	if (ordered)
	{   printf ("%7ld %s", order, q);
	    order ++;
	}
	else
	    printf ("%7ld %s", order, p);
    }

    return 0;
}
 


src/percent.c
/* 
 * $Id$
 * Bozomake Copyright (C) 1997,2002,2003 Cristelle Barillon & Jean-Daniel Pauget
 * A whole set of building utilities
 *
 * bozomake@disjunkt.com  -  http://bozomake.disjunkt.com/
 * 
 * This file is part of Bozomake.
 * 
 * Loopsoids is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Loopsoids is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Loopsoids; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * you can also try the web at http://www.gnu.org/
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct entry
{	char * ligne;
	long num;
	struct entry * next;
};

int compare_entry (const void *a, const void *b)
{	return (*(struct entry * const *)b)->num - (*(struct entry * const *)a)->num;
}

int main (int nb, char ** cmde)
{
	char buf [4096];
	
	int order = 1;

	struct entry 	*first, 
			*current,
			**tab,
			**last = &first;

	long  tot = 0,
		n = 0,
		i;

	for (i=1 ; i<nb ; i++) {
	    if (cmde[i][0] == '-') {
		switch (cmde[i][1]) {
		    case 'r':
			order = 0;
			break;
		    default:
			break;
		}
	    }
	}

	while (fgets (buf, 4096, stdin) != NULL)
	{
		if ( (*last = (struct entry *) malloc (sizeof (struct entry))) == NULL )
		{	fprintf (stderr, "un malloc a retourne NULL : j'abrege ! \n");
			return 1;
		}
		if ( ((*last)->ligne = (char *) malloc (strlen (buf)+1)) == NULL )
		{	fprintf (stderr, "un malloc a retourne NULL : j'abrege ! \n");
			return 1;
		}
		strcpy ((*last)->ligne, buf);
		(*last)->num = atol (buf);
		tot += (*last)->num;
		n++;
		last = &((*last)->next);
	}

	*last = NULL;

	if (order)
	{	tab = (struct entry **) malloc (n * sizeof (struct entry *));	
		if (tab == NULL)
		{	fprintf (stderr, "ca semble incroyable mais je n'ai pas assez de memoire pour trier...\n");
			order = 0;
		}
		i = 0;
	}

	if (order)
	{	if (tot != 0)
			printf ("        %3ld %% %7ld %s (%ld items)\n", (100 * tot)/tot, tot, "TOTAL", n);
		current = first;
		while (current != NULL)
		{	tab [i] = current;
			i ++;
			current = current-> next;
		}

		qsort (tab, n, sizeof (struct entry *), compare_entry);

		for (i=0 ; i<n ; i++)
			printf ("%7ld %3ld %% %s", i+1, (100 * tab[i]->num)/tot, tab[i]->ligne);
	}
	else
	{	if (tot != 0)
			printf ("        %3ld %% %7ld %s (%ld items)\n", (100 * tot)/tot, tot, "TOTAL", n);
		current = first;
		i = 1;
		while (current != NULL)
		{	printf ("%7ld %3ld %% %s", i, (100 * current->num)/tot, current->ligne);
			current = current-> next;
			i++;
		}	
	}

	return 0;
}

 


src/revtail.c
/* 
 * $Id$
 * Bozomake Copyright (C) 2002,2003 Cristelle Barillon & Jean-Daniel Pauget
 * A whole set of building utilities
 *
 * bozomake@disjunkt.com  -  http://bozomake.disjunkt.com/
 * 
 * This file is part of Bozomake.
 * 
 * Loopsoids is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Loopsoids is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Loopsoids; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * you can also try the web at http://www.gnu.org/
 */
#define _LARGEFILE64_SOURCE

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#define BSIZE 1024

#ifdef USELSEEK64

#define BIGOFF_T off64_t
#define BIGLSEEK lseek64
	/* JDJDJDJD one should check this one is bound to lseek64 ??? */
#define BIGO_LARGEFILE O_LARGEFILE

#else	/* ios misses lseek64 or equivalent ??? */

#define BIGOFF_T off_t
#define BIGLSEEK lseek
#define BIGO_LARGEFILE 0

#endif	/* USELSEEK64 */

typedef struct STFlushLBuf {
    struct STFlushLBuf * next,
		       * prev;
    char buf[BSIZE];
    int cur;
    int nbeol;
} FlushLBuf;

int nmalloc = 0;
int nfree = 0;
long long nblignes = -1;

void freeallbuf (FlushLBuf *b) {
    FlushLBuf *next;
    while (b != NULL) {
	next = b->next;
	free (b);
	b = next;
	nfree ++;
    }
}

void flush_from_end (FlushLBuf *b);

void flush (FlushLBuf *b, int p) {
    while (b != NULL) {
	FlushLBuf *temp = b->next;
	int size = b->cur - p;
	int size_out = 0;

	while (size > 0) {
	    ssize_t l = write (1, b->buf+p+size_out, size);
	    if (l != -1) {
		size_out += l;
		size -= l;
	    } else {
		fprintf (stderr, "could not output %d byte ??\n", size);
		return;
	    }
	}
	if (p==0) {
	    free (b);
	    nfree ++;
	}
	    
	b = temp;
	p = 0;
    }
}

void buftailrev (int fd) {
    size_t nbread = 0;
    long long toteol = 0;
    FlushLBuf *first = NULL,
	      *prev = NULL,
	      *b = NULL;

    do {
	int i;

	b = (FlushLBuf *) malloc (sizeof(FlushLBuf));
	if (b == NULL) {
	    fprintf (stderr, "could not allocate about %d bytes ??\n", (int)sizeof(FlushLBuf));
	    freeallbuf (first);
	    return;
	}
	nmalloc ++;
	b->nbeol = 0;
	b->next = NULL;
	if (first == NULL)
	    first = b;
	if (prev != NULL) {
	    prev->next = b;
	}
	b->prev = prev;

	nbread = read (fd, b->buf, BSIZE);
	b->cur = nbread;

	for (i=0 ; i<nbread ; i++)
	    if ((b->buf[i] == 10) || (b->buf[i] == 13))
		b->nbeol ++;
	
	toteol += b->nbeol;

	if (nblignes > 0) {
	    while ((first != NULL) && (toteol - first->nbeol > nblignes)) {
		FlushLBuf *temp = first;
		toteol -= first->nbeol;
		first = first->next;
		free (temp);
		nfree ++;
	    }
	}

	prev = b;
    } while (nbread > 0);

    flush_from_end (b);

    // freeallbuf (first);
}

void tailrev (int fd) {
    int i;
    long long lignesflushed = 0;
    FlushLBuf *next = NULL;
    BIGOFF_T o = BIGLSEEK (fd, 0, SEEK_END);
    if (o == (BIGOFF_T)-1) {
	fprintf (stderr, "could not seek to end of file ??\n");
	return;
    }
    while (o > 0) {
	int l;
	BIGOFF_T size_to_read;

	FlushLBuf *b = (FlushLBuf *) malloc (sizeof(FlushLBuf));
	if (b == NULL) {
	    fprintf (stderr, "could not allocate about %d bytes ??\n", (int)sizeof(FlushLBuf));
	    freeallbuf (next);
	    return;
	}
	nmalloc ++;
	b->next = next;
	if (o < BSIZE)
	    size_to_read = o;
	else
	    size_to_read = BSIZE;
	b->cur = size_to_read;

	o = BIGLSEEK (fd, -size_to_read, SEEK_CUR);
	if (o == (BIGOFF_T)-1) {
	    fprintf (stderr, "could not seek back %d bytes in file ??\n", (int)size_to_read);
	    freeallbuf (b);
	    return;
	}

	l = read (fd, (void *)(b->buf), size_to_read);
	if (l != size_to_read) {
	    fprintf (stderr, "could not read %d bytes in file ??\n", (int)size_to_read);
	    freeallbuf (b);
	    return;
	}

	o = BIGLSEEK (fd, -size_to_read, SEEK_CUR);
	if (o == (BIGOFF_T)-1) {
	    fprintf (stderr, "could not seek back %d bytes in file ??\n", (int)size_to_read);
	    freeallbuf (b);
	    return;
	}

	next = b;
	for (i=size_to_read-1 ; i>=0 ; i--) {
	    if ((b->buf[i] == 10) || (b->buf[i] == 13)) {
		if (nblignes > 0) {
		    if (lignesflushed <= nblignes) {
			lignesflushed ++;
			flush (b, i+1);
		    } else {
			freeallbuf (b);
			return;
		    }
		} else
		    flush (b, i+1);
		b->next = NULL;
		b->cur = i+1;
	    }
	}

	if (o == 0) flush (b, 0);
    }
}

void flush_from_end (FlushLBuf *b)
{
    int i;
    long long lignesflushed = 0;

    while (b != NULL) {
	for (i=b->cur-1 ; i>=0 ; i--) {
	    if ((b->buf[i] == 10) || (b->buf[i] == 13)) {
		if (nblignes > 0) {
		    if (lignesflushed <= nblignes) {
			lignesflushed ++;
			flush (b, i+1);
		    } else {
			return;
		    }
		} else
		    flush (b, i+1);
		b->next = NULL;
		b->cur = i+1;
	    }
	}
	if (b->prev == NULL) flush (b, 0);
	b = b->prev;
    }
}


int main (int nb, char ** cmde) {
    int i;
    int stillarg = 1;
    int nbfiles = 0;

    for (i=1 ; i<nb ; i++) {
	if (stillarg) {
	    if (cmde[i][0] == '-') {
		if (strcmp ("--", cmde[i]) == 0) {
		    stillarg = 0;
		    continue;
		} else {
		    nblignes = atoll (cmde[1] + 1);
		    continue;
		}
	    }
	}

	nbfiles ++;
	int fd = open (cmde[i], BIGO_LARGEFILE);
	if (fd != -1) {
	    tailrev (fd);
	    close (fd);
	} else {
	    fprintf (stderr, "could not open %s for reading\n", cmde[i]);
	}
    }
    if (nbfiles == 0) {
	buftailrev (0);
    }

//    fprintf (stderr, "malloc = %d, free = %d\n", nmalloc, nfree);

    return 0;
}



src/scramble.c
/* 
 * $Id$
 * Bozomake Copyright (C) 2002,2003 Cristelle Barillon & Jean-Daniel Pauget
 * A whole set of building utilities
 *
 * bozomake@disjunkt.com  -  http://bozomake.disjunkt.com/
 * 
 * This file is part of Bozomake.
 * 
 * Loopsoids is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Loopsoids is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Loopsoids; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * you can also try the web at http://www.gnu.org/
 */
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>

#define MABIGLINE 4096
#define MAXNB 4096

char *lig[MAXNB];
int o[MAXNB];

int
main (void)
{
  char buf[MABIGLINE];
  int i, n, raison = 1;

  srand ((unsigned int) time (NULL));

  while (raison == 1)
    {
      raison = 0;
      n = 0;

      while ((n < MAXNB) && (fgets (buf, MABIGLINE, stdin) != NULL))
	{
	  if ((lig[n] = (char *) malloc (strlen (buf) + 1)) != NULL)
	    {
	      strcpy (lig[n], buf);
	      n++;
	    }
	  else
	    {
	      raison = 1;
	      break;
	    }
	}

      for (i = 0; i < n; i++)
	o[i] = i;

      for (i = 0; i < n; i++)
	{
	  int temp, s;

	  s = (int) ((double) (n - 1) * rand () / (RAND_MAX + 1.0));
	  temp = o[i];
	  o[i] = o[s];
	  o[s] = temp;
	}
      for (i = 0; i < n; i++)
	{
	  fputs (lig[o[i]], stdout);
	  free (lig[o[i]]);
	}
    }
  return 0;
}


src/countsort.c
/* 
 * $Id$
 * Bozomake Copyright (C) 1997,2002,2003 Cristelle Barillon & Jean-Daniel Pauget
 * A whole set of building utilities
 *
 * bozomake@disjunkt.com  -  http://bozomake.disjunkt.com/
 * 
 * This file is part of Bozomake.
 * 
 * Loopsoids is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Loopsoids is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Loopsoids; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * you can also try the web at http://www.gnu.org/
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main (int nb, char ** argv)
{
	long count = 0, order = 0;
	int ordered = 0;
	char buf1 [4096],
	     buf2 [4096],
	     *p, *q, *r;
	p = buf1, q = buf2;
	*p = 0;

	{	int i;
		for (i=1 ; i<nb ; i++)
		{	if (*argv[i] == '-')
			switch (argv[i][1])
			{	case 'n':
				case 'N':
					ordered = 1;
					break;
			}
		}
	}

	while (fgets (q, 4096, stdin) != NULL)
	{	if (strcmp (q,p) != 0)
		{	
			if (count) 
			{	if (ordered)
				{	printf ("%7ld %7ld %s", order, count, p);
					order ++;
				}
				else
					printf ("%7ld %s", count, p);
			}
			r = p;
			p = q;
			q = r;
		
			count = 1;
		}
		else
			count ++;
	}
	if (count) 
	{	if (ordered)
		{	printf ("%7ld %7ld %s", order, count, p);
			order ++;
		}
		else
			printf ("%7ld %s", count, p);
	}

	return 0;
}
 


src/brev.c
/*
 *  $Id$
 * 
 *  brev perfoms a left to righ conversion of a text file.
 *  it is intended to supply the 'rev' functionnality when it is missing,
 *  though it doesn't support multi-bytes chars nor locales, and probably more.
 *
 *  jd - 27-04-2004 jdrev@disjunkt.com
 *
 */
/* 
 * $Id$
 * Bozomake Copyright (C) 2002,2003 Cristelle Barillon & Jean-Daniel Pauget
 * A whole set of building utilities
 *
 * bozomake@disjunkt.com  -  http://bozomake.disjunkt.com/
 * 
 * This file is part of Bozomake.
 * 
 * Loopsoids is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Loopsoids is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Loopsoids; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * you can also try the web at http://www.gnu.org/
 */


#include <stdio.h>
#include <stdlib.h>

typedef struct STFlushBuf {
    int first;
    int last;
    struct STFlushBuf * previous;
    char buf[1024];
} FlushBuf;

void flushit (FlushBuf * fb)
{
    int k;
    FlushBuf * previous;
    if (fb == NULL) {
	fprintf (stderr, "WARNING: flushit got a null FlushBuf\n");
	return;
    }
    // JDJDJDJD the following is uggly, a left to right conversion
    // of the buffer should be performed, then a BLOCK write, instead of this crap
    for (k=fb->last ; k>=fb->first ; k--)
	printf ("%c", fb->buf[k]);
    // end of crap
    previous = fb->previous;
    free (fb);
    if (previous != NULL)
	flushit (previous);
}

void brev (FILE * f,FlushBuf * previous)
{
    FlushBuf * cur;
    char *buf, *p;
    int nbread, i, j;

    cur = (FlushBuf *) malloc (sizeof (FlushBuf));
    if (cur == NULL) {
	fprintf (stderr, "brev: ERROR: could not allocate 1024 bytes\n");
	return;
    }

    buf = &cur->buf[0];
    
    nbread = fread (buf, 1, 1024, f);
    i=0;
    p=&buf[0];
    while (i<nbread) {
	j=i;
	while ((i<nbread) && (*p!=10) && (*p!=13)) i++, p++;
	if (i == nbread) {
	    cur->first=j;
	    cur->last=i-1;
	    cur->previous = previous;
	    brev (f, cur);
	} else {
	    int k;
    // JDJDJDJD the following is uggly, a left to right conversion
    // of the buffer should be performed, then a BLOCK write, instead of this crap
	    for (k=i-1 ; k>=j ; k--)
		printf ("%c", buf[k]);
    // end of crap
	    if (previous != NULL) {
		flushit (previous);
		previous = NULL;
	    }
	    printf ("%c", *p);
	    i++, p++;
	    if (i == nbread) {
		free (cur);
		brev (f, NULL);
	    }
	}
    }
}

int main (int nb, char ** cmde)
{
    int i;

    if (nb == 1)
	brev (stdin, NULL);
    else for (i=1 ; i<nb ; i++) {
	FILE *f = fopen (cmde[i], "r");
	if (f != NULL) {
	    brev (f, NULL);
	    fclose (f);
	} else {
	    fprintf (stderr, "could not open %s for reading\n", cmde[i]);
	}
    }
    return 0;
}

